### ISSUE TITLE
'Development' returned as environment name in Azure after disabling any function

### ISSUE THREAD

**Issue by Ads5555** [Azure/azure-functions-host#6239]
#### Investigative information

- Timestamp: 2020-06-18T03:00:26Z
- Function App version: v3
- Function App name: FunctionApp220200618084855
- Function name(s) (as appropriate): EnvironmentTestFunction
- Invocation ID:
- Region: Australia Southeast

#### Repro steps

1. Create a new default Azure Function solution.
2. Add a function that simply prints out the hosting environment name via the log property.
    - `_hostingEnvironment.EnvironmentName`
3. Deploy to a new Azure instance and view the output of running the function via 'Code + Test' interface.
4. Note that it states the environment is 'Development'. 
    - [Docs](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings) state the default should be Prod in Azure. This appears to be another issue?
5. Set `AZURE_FUNCTIONS_ENVIRONMENT` to a value in the configuration (e.g. 'Staging').
6. Re-run the function and note its returned correctly as 'Staging'.
7. Disable the function and then re-enable.
8 Re-run the function and note that it now says 'Development'. Restarting the functions doesn't resolve the issue. `AZURE_FUNCTIONS_ENVIRONMENT` is disregarded.

#### Expected behavior

- Disabling and enabling functions should not change if the `AZURE_FUNCTIONS_ENVIRONMENT` is used or not.

#### Actual behavior

- `AZURE_FUNCTIONS_ENVIRONMENT` is ignored after disabling a function and is useless for determining the environment correctly.
- I havent found a way to fix this after it becomes broken.

#### Known workarounds

Use your own environment key to pull from the configuration.

#### Related information 

<details>
<summary>Source</summary>

```csharp
public class EnvironmentTestFunction
{
    private readonly IHostingEnvironment _hostingEnvironment;

    public EnvironmentTestFunction(IHostingEnvironment hostingEnvironment)
    {
        _hostingEnvironment = hostingEnvironment;
    }

    [FunctionName(nameof(EnvironmentTestFunction))]
    public void EnvironmentTest([TimerTrigger("0 */1 * * * *", RunOnStartup = true)]TimerInfo myTimer, ILogger log)
    {
        try
        {
            log.LogWarning("Finding Env");
            log.LogWarning($"_hostingEnvironment.EnvironmentName: {_hostingEnvironment.EnvironmentName}");
            log.LogWarning($"_hostingEnvironment.IsProduction: {_hostingEnvironment.IsProduction()}");
        }
        catch (Exception e)
        {
            log.LogError(e, $"Exception caught {e.Message}");
        }
    }
}

```
</details>

**Comment by devdorn** [comment]
I've the same Problem: since 2020-06-18 the default Environment on Azure Functions (West Europe, Runtime Version 3.0.13901.0) has changed from "Production" to "Development".  
I've never disabled my Function, but I had an Deployment on 2020-06-18.

**Comment by IGx89** [comment]
I'm seeing some very weird behavior with environment name as well. My function didn't have AZURE_FUNCTIONS_ENVIRONMENT set and was returning "Production" in Azure, but when I did set AZURE_FUNCTIONS_ENVIRONMENT it changed to "Development" and now no matter what I set AZURE_FUNCTIONS_ENVIRONMENT to (or even remove it) it always is "Development" now. Something definitely seems buggy in that area.

**Comment by wagich** [comment]
I seem to have the same problem. No matter what value I set for `AZURE_FUNCTIONS_ENVIRONMENT`, `IHostEnvironment.EnvironmentName` always returns "Development".

**Comment by smokedlinq** [comment]
This happened to me today, nothing I could do would change `IHostEnvironment.EnvironmentName` to what I set `AZURE_FUNCTIONS_ENVIRONMENT` value as in the configuration. I opted for a code-fix that checks the environment variable first, and if it's empty defaults to the `IHostEnvironment` value. Locally, when debugging with v3.0.2630 core tools, this does not happen.

**Comment by madelinegordon** [comment]
Hello, we will investigate this issue and provide information based on our findings. Thank you.

**Comment by rigdern** [comment]
Is this related to #6085?

**Comment by Ads5555** [comment]
> Is this related to #6085?

My functions were respecting the prod environment config until I disabled them. It could be related but I think the investigation would need to determine that.

**Comment by wagich** [comment]
In my case it happend from the beginning on a completely new function app I deployed to via docker. It might well be two separate or related problems and me hitting #6085 too…

**Comment by madelinegordon** [comment]
Hello, we were able to repro this issue, and we believe the cause is related to the scoping of components within the Functions Host - the WebHost (outer scope) component is only fully restarted when a Function app is stopped and then started up again, whereas the ScriptHost (inner scope) is the only component that is restarted when a Function app is restarted from the portal - essentially we only do a partial restart in this case. It seems that whenever only a partial restart occurs, 'IHostEnvironment.EnvironmentName` becomes 'Development'. This can be mitigated by stopping your Function app, waiting a few seconds, and then starting it back up again. 

Moving this to Triaged.

**Comment by madelinegordon** [comment]
Hello, we were able to repro this issue, and we believe the cause is related to the scoping of components within the Functions Host - the WebHost (outer scope) component is only fully restarted when a Function app is stopped and then started up again, whereas the ScriptHost (inner scope) is the only component that is restarted when a Function app is restarted from the portal - essentially we only do a partial restart in this case. It seems that whenever only a partial restart occurs, 'IHostEnvironment.EnvironmentName` becomes 'Development'. This can be mitigated by stopping your Function app, waiting a few seconds, and then starting it back up again.

**Comment by ericsampson** [comment]
@madelinegordon @ankitkumarr is this scheduled as a priority fix? This seems like a bad bug - as mentioned in #6085 , it's happens not only in the relatively rare case of disabling/enabling a function, but also after deploys or restarts.
If this was the same type of bug in plain App Service where customer's code was running with the wrong environment settings, there would be a big freakout and fast fix - this can cause bad things like customers calling the dev APIs instead of prod APIs and subsequent data/PII issues...

**Comment by ankitkumarr** [comment]
@fabiocav and @anthonychu -- Thoughts on the investigation and priority for this?

**Comment by pragnagopa** [comment]
Assigning to Sprint 87 as a stretch goal

**Comment by fabiocav** [comment]
This might slip into the next sprint. Updating to reflect that.

**Comment by ITRON-kpalmberg** [comment]
I am having similar issues. I am unable to change it to anything but "Development" when deployed in Azure. According to the docs, it should have defaulted to "Production" in Azure anyways. I set "AZURE_FUNCTIONS_ENVIRONMENT" directly in the function app configuration. I also tried "ASPNETCORE_ENVIRONMENT" for kicks, even though the docs said for function apps you should use "AZURE_FUNCTIONS_ENVIRONMENT". I was unable to get it to change in Azure. I'm going to have to use something else to determine environment such as a completely separate environmental variable. Blew a couple hours debugging this thinking I was making mistakes which is really unfortunate. Hopefully this can be resolved soon.

FYI: This did not occur after I disabled/enabled a function as I have never disabled any of my functions. I only started running into this issue after I had a need to determine environment when I added appsettings.json files (i.e appsettings.Staging.json).

**Comment by BloodSeaker** [comment]
> I am having similar issues. I am unable to change it to anything but "Development" when deployed in Azure. According to the docs, it should have defaulted to "Production" in Azure anyways. I set "AZURE_FUNCTIONS_ENVIRONMENT" directly in the function app configuration. I also tried "ASPNETCORE_ENVIRONMENT" for kicks, even though the docs said for function apps you should use "AZURE_FUNCTIONS_ENVIRONMENT". I was unable to get it to change in Azure. I'm going to have to use something else to determine environment such as a completely separate environmental variable. Blew a couple hours debugging this thinking I was making mistakes which is really unfortunate. Hopefully this can be resolved soon.
> 
> FYI: This did not occur after I disabled/enabled a function as I have never disabled any of my functions. I only started running into this issue after I had a need to determine environment when I added appsettings.json files (i.e appsettings.Staging.json).

We experience the same issue. Since we have no way to update the environment to anything else, it causes a separate 'environment' settings file to load, which overwrites some settings from the regular settings file. Therefore, we cannot deploy our solution to our validation environment, as this one also returns 'Development' from the context.EnvironmentName property.

Why is this ticket open for several months now, as this is causing actual issues and not just a gimmick which is failing?

**Comment by ericsampson** [comment]
It is really disappointing that this continues to slip. It is a bad bug

**Comment by ITRON-kpalmberg** [comment]
@BloodSeaker It's not optimal or elegant but you can work around the issue. You will just have to manage the environment setting yourself, such as by your own environmental variable, and not rely on the hosting environment property and rely on your own implementation instead. Then you can use your own custom app settings (probably an environmental variable) to determine which app settings files to load.

**Comment by baharedankoyuncu** [comment]
We experienced this as well but were able to work around it both locally and in Azure. We are setting the `AZURE_FUNCTIONS_ENVIRONMENT` environment variable even though it was mentioned in previous comments that it didn't work.

**Comment by fabiocav** [comment]
@ericsampson @baharedankoyuncu are you still able to repro this? If so, can you please provide information about the version you're using and your setup? 
Thanks!

**Comment by ericsampson** [comment]
@fabiocav are you asking because something has changed?
We were using V2. I'd have to back out the workaround that we ended up using, in order to verify that it's still happening. I'm not sure that I'll have time to do that anytime soon.

**Comment by fabiocav** [comment]
@ericsampson it's just not something our team has been able to repro using the latest bits, so getting confirmation on the version would be helpful.

Adding @ankitkumarr and @pragnagopa if additional context is needed.

**Comment by pragnagopa** [comment]
@ericsampson - We spent sometime going over the repro steps mentioned in the description.  We could not repro the issue with the latest runtime. Tried few combinations. It would be great if you can try to repro on a different app to ensure your prod app is not effected.

**Comment by devdorn** [comment]
![image](https://user-images.githubusercontent.com/399354/101877641-9b30e600-3b8e-11eb-8d0f-25bbe1e84c48.png)

![image](https://user-images.githubusercontent.com/399354/101877347-25c51580-3b8e-11eb-939b-f02aa89a3911.png)

I've implemented an TimerTrigger function  which runs daily and prints out the Env setting.

This was yesterday (triggered):
![image](https://user-images.githubusercontent.com/399354/101877745-cc111b00-3b8e-11eb-9cf6-02ea1c3c7f41.png)

This was today, before I triggered the function manually over the portal > RunOnStartup:
![image](https://user-images.githubusercontent.com/399354/101877026-acc5be00-3b8d-11eb-941b-6c1ae9fd35f9.png)

Run in Portal:
![image](https://user-images.githubusercontent.com/399354/101877177-e4346a80-3b8d-11eb-8181-e9707cde3464.png)

And this was an triggered execution on Dec 4th:
![image](https://user-images.githubusercontent.com/399354/101877279-0a5a0a80-3b8e-11eb-9f56-d92df00fdec9.png)

For me it's an unreliable behavior.

**Comment by ericsampson** [comment]
It was unreliable for us as well, it didn’t happen every time. I’m glad you have a way to reproduce it :)

**Comment by ericsampson** [comment]
FWIW in our case, we had set an explicit portal setting value of either Staging or Production and were seeing it return Dev sometimes after lifecycle events. Doing a full stop and restart, as was suggested upthread, did not help.

**Comment by Simon-Gregory-LG** [comment]
Any update on this issue? 

I've encountered this nasty bug in the last few days, initially I thought it was related to the Snapshot Debugger, but seems I'm getting the same inconsistent behaviour with Environment.

I've found that the function can be honouring the setting one minute, but then either changing a AppService configuration setting in the Azure Portal, performing an ARM deploy or deploying the Function app itself can put it into this state. Only 'fix' seems to be to shutdown the function for a time and then start it. 

This is clearly not useful for a production scenario (particularly as it's reverting to 'Development'!), and as others have said should be prioritised given that it can cause significant behaviour change for seemingly no obvious reasons.

**Comment by Simon-Gregory-LG** [comment]
As stopping and starting the app is never an acceptable or workable solution, this is my temporary code workaround for `FunctionsStartup.ConfigureAppConfiguration`:

``` csharp
        public override void ConfigureAppConfiguration(IFunctionsConfigurationBuilder builder)
        {
            var context = builder.GetContext();

            // Workaround for unstable EnvironmentName in Azure 
            // (see https://github.com/Azure/azure-functions-host/issues/6239)
            var environmentName = Environment.GetEnvironmentVariable("AZURE_FUNCTIONS_ENVIRONMENT") ??
                                    Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ??
                                    context.EnvironmentName;

            builder.ConfigurationBuilder
                        .SetBasePath(context.ApplicationRootPath)
                        .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                        .AddJsonFile($"appsettings.{environmentName}.json", optional: true, reloadOnChange: true)
                        .AddEnvironmentVariables();
        }
```

But please, please, please fix this. I've **lost many hours** to this already - others will surely, or even worse may not even notice they have incorrect configurations running.

**Comment by devdorn** [comment]
Maybe helpful, but I do encounter this behavior only on Functions not on AppServices.

**Comment by fabiocav** [comment]
> Maybe helpful, but I do encounter this behavior only on Functions not on AppServices.

@devdorn can you clarify what you mean by that? 

Thanks for the additional details. What you've share points to a possible race, so that narrows things down a bit.

**Comment by ericsampson** [comment]
I'm guessing Consumption plan vs fixed plan.

**Comment by fabiocav** [comment]
I see. Thank you! That does provide some additional clues about what might be happening here. Thank you!

**Comment by BloodSeaker** [comment]
We are running on app service plan and not consumption and experience the same issue.
We now set it via ARM deploy to our own custom value just to be sure. That seems to hold, but should not be there and one should be able to rely on it sticking to 'production'
________________________________

**Comment by devdorn** [comment]
> > Maybe helpful, but I do encounter this behavior only on Functions not on AppServices.
> 
> @devdorn can you clarify what you mean by that?
> 
> Thanks for the additional details. What you've share points to a possible race, so that narrows things down a bit.

What I was saying is, that I encounter this behavior only on Azure Serverless Functions and not on Azure Web Apps or how they are called. I  call them AppServices. The thing is, they are similar with Functions in terns of AppSettings or Configuration. Only there have more Web Features. But the Settings Part and the way environment variables get injected is the same. There are in many ways identical and yet only the functions framework has this issue.
I use only App Service Plans because of having to use VNet to get OnPrem connections. I do not use Consumption Plans on my Functions.

**Comment by helmerm** [comment]
Encountered issue on C# Function v3 hosted in App Service Plan.

- Using code as [documented](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection#customizing-configuration-sources)
- AZURE_FUNCTIONS_ENVIRONMENT set to Test
- verified app settings and environment variables of process (via kudu) --> show Test
- settings from appsettings.Development.json are used

Edit: We use WEBSITE_RUN_FROM_PACKAGE for the deployment of the function app.

**Comment by ericsampson** [comment]
@fabiocav any progress on this? It’s not good. Thanks!

**Comment by fabiocav** [comment]
@ericsampson unfortunately, this was an issue that we were unable to look at this sprint due to competing items. We're bumping this up in priority and we'll provide an update as soon as we have something to report. 

Moving to sprint 95 so we can continue to track this there.

**Comment by fabiocav** [comment]
@ericsampson apologies for the delay on this, but with the current priorities, we haven't been able to look at this issue yet. Assigning to sprint 97 as that more appropriately reflects when this should be done.

**Comment by jonsagara** [comment]
I'm yet another party affected by this.

I just created four new function apps, and it's happening in all four.

Runtime version 3
Consumption Plan
.NET Core SDK 3.1.406
WEBSITE_RUN_FROM_PACKAGE = 1
Deployed via Azure DevOps Pipelines
AZURE_FUNCTIONS_ENVIRONMENT = Staging (or PreProduction, or Production. None work. It always returns Development.)

Looking forward to the fix. Thank you.

ETA: I confirmed that a hard stop, followed by a 10 count, followed by a restart causes the function to pick up the expected `AZURE_FUNCTIONS_ENVIRONMENT` value from configuration.

**Comment by betim-raqi** [comment]
Very unstable behaviour.

With one function app I created I could sometimes make it to change the environment to Production, that after stopping the application and then starting it again.

However, now one hour later I created a new function and can not seem to change the environment whatsoever, it always says Development (even though in the documentation it says it should default to production)

**Comment by Simon-Gregory-LG** [comment]
I've commented before on this one and even [come up with a workaround](https://github.com/Azure/azure-functions-host/issues/6239#issuecomment-752757645), but this ticket seems to keep getting pushed back, so feel the need to re-echo the need for attention here. 

The workaround is ok if you know you have the problem, but there will be plenty of people who aren't even aware this is happening. While I take small comfort that I'm not in that group, I shudder to think what could be happening elsewhere with this issue.

I struggle to think of a more critical problem than a system intermittently and indiscriminately loading its Development configuration while in Production - yet we are still not even past the `needs-investigation` phase here.

**Comment by krosk93** [comment]
Hello, any updates? We are experiencing right now the same behaviour. It suddenly changed from Production to Development environment. Thanks.

**Comment by fabiocav** [comment]
@krosk93 there hasn't been a lot of progress on this aside from additional investigation and identification of a couple of additional scenarios.

One thing that is worth noting, as that is what was causing the behavior for some customers, is that using any of the `admin` APIs (e.g. opening the portal or working with the hosted version directly or with one of the tools) **will place the host in debut mode for a period of time**. This enables some debug services used by the portal, like debug logging. Is it possible this is what is impacting you?

**Comment by ericsampson** [comment]
@fabiocav could you expand what that means?

I don't believe that is the only scenario we've experienced, as the issue was initially discovered after an application deploy when we noticed from logging to our ES cluster that it was calling APIs in the wrong environment.

**Comment by DavidZidar** [comment]
I pointed out the offending code in my bug report #6085 nearly a year go and the code is still there.

`EnableDebugModeFilter` is used by `HostController` when interacting with features in the portal:
https://github.com/Azure/azure-functions-host/blob/9c2ce4c108b08724b623df088d9ed4d232af116a/src/WebJobs.Script.WebHost/Controllers/HostController.cs#L64-L65

`EnableDebugModeFilter` flags that debug mode should be used:
https://github.com/Azure/azure-functions-host/blob/9c2ce4c108b08724b623df088d9ed4d232af116a/src/WebJobs.Script.WebHost/Filters/EnableDebugModeFilter.cs#L26-L29

The host then restarts and the builder sets the environment to `Development`:
https://github.com/Azure/azure-functions-host/blob/9c2ce4c108b08724b623df088d9ed4d232af116a/src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs#L140-L144

Please find another way to enable debug services, DO NOT change the environment like this, it breaks all sorts of things.

**Comment by fabiocav** [comment]
@ericsampson this is not the only root cause, but wanted to make sure it was noted here as it is what impacts a lot of the apps running into this.

Right, @DavidZidar that is the logic that will be addressed.

**Comment by tiwaripradeep123** [comment]
I see intermittent issue too and today all the functions are failing because they are using wrong settings file in production.

**Comment by rmaziarka** [comment]
I experienced the same problem today. It's great that there is a @Simon-Gregory-LG workaround but such behavior of AF makes us spend hours / days on solving the wrong types of problems.

**Comment by bedwards-humana** [comment]
I'm having the same issue.  Will work around for now by retrieving the AZURE_FUNCTIONS_ENVIRONMENT variable directly instead of relying on builder.GetContext().EnvironmentName.

**Comment by ericsampson** [comment]
Hi @fabiocav, I saw your PR #7304 , glad to see some movement. However, I believe that this is one of those rare occasions where a breaking change _should_ be made, and then people can opt into the current behavior if that's what they actually want.

Also, is there going to be continued investigation as to why this issue occurs unrelated to debug mode? I don't want the issue to stop being investigated just because one way that it gets triggered was addressed.

Thanks!

**Comment by fabiocav** [comment]
@ericsampson there will be further investigation here, but we didn't want to block on that in order to address a known sceanario.

I do understand where you're coming from with the breaking change, and we did discuss this, but with the broad number of scenarios we see on Functions, we did have concrete examples and evidence of customers who would be negatively impacted by the change if that was made the default, but the new behavior will become the default with the next major version being released later this year.

**Comment by ericsampson** [comment]
Thanks @fabiocav, I figured it wasn't the end of the road but wanted to verify.

Re the second part, I hear what your saying... But most of the people on this issue would probably see the original behavior as a `bug`, or an misdesigned implementation at the least. Could you help me understand the scenario in which the current behavior is desirable?

**Comment by ericsampson** [comment]
I'm not trying to be a pain in the arse, I really appreciate the progress here.

Just trying to make sure the widest number of people are prevented from having to go through the pain that the people in this ticket have. Never mind everyone affected but they didn't/haven't even realize it.

**Comment by fabiocav** [comment]
@ericsampson no worries, not being a pain at all, that is our desire as well. But we do have evidence and concrete examples of applications that were relying on the "buggy behavior" and relying on the environment name change when portal/API interactions were taking place. Our tolerance for breaking changes is VERY low, so even in cases like this, we have to be safe with the approach. That, combined with the fact that our next major is just months away, impacted the decision.

**Comment by ericsampson** [comment]
I understand, thanks Fabio. Cheers

**Comment by ericsampson** [comment]
One quick question Fabio; when you exit the "Code and Test" feature in the blade, is the host being reset to the original environment?

**Comment by fabiocav** [comment]
> One quick question Fabio; when you exit the "Code and Test" feature in the blade, is the host being reset to the original environment?

Not immediately. The state is eventually reset once debug times out (need to double check, but I believe that happens after 15 mins)

**Comment by psddp** [comment]
Looks like it was a design decision to always have "Development" when running local, see https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azure_functions_environment

![image](https://user-images.githubusercontent.com/7760275/120215434-b9a1a100-c203-11eb-855b-201fc8668db9.png)

**Comment by unruledboy** [comment]
This is very annoying. Finally we had to directly read the environment value to determine what appsettings.json will be loaded:

```
        public override void ConfigureAppConfiguration(IFunctionsConfigurationBuilder builder)
        {
            var hostBuilderContext = builder.GetContext();
            var isDevelopmentEnvironment = Environment.GetEnvironmentVariable("AZURE_FUNCTIONS_ENVIRONMENT").ToLower().Contains("dev");
           //your usage of isDevelopmentEnvironment here
        }  
```

**Comment by UsmanMohammad** [comment]
> Looks like it was a design decision to always have "Development" when running local, see https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azure_functions_environment
> 
> ![image](https://user-images.githubusercontent.com/7760275/120215434-b9a1a100-c203-11eb-855b-201fc8668db9.png)

It's quite short-sighted for them to assume only `Development`, `Staging` and `Production` are the only environment names used by people.

What I've done is to just ignore "AZURE_FUNCTIONS_ENVIRONMENT" and built my configuration this way


        private IConfiguration _configuration;
        private string _environmentName;
        public override void ConfigureAppConfiguration(IFunctionsConfigurationBuilder builder)
        {
            _environmentName = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
            
            var applicationRootPath = builder.GetContext().ApplicationRootPath;
            builder.ConfigurationBuilder
                .AddJsonFile(Path.Combine(applicationRootPath, "appsettings.json"), optional: true, reloadOnChange: false)
                .AddJsonFile(Path.Combine(applicationRootPath, $"appsettings.{ _environmentName}.json"), optional: true, reloadOnChange: false)
                .AddEnvironmentVariables();

            if (!string.Equals(_environmentName, "Development", StringComparison.OrdinalIgnoreCase))
            {
                var builtConfig = builder.ConfigurationBuilder.Build();
                var keyVaultName = builtConfig["KeyVaultName"];
                if (!string.IsNullOrWhiteSpace(keyVaultName))
                {
                    var secretClient = new SecretClient(
                        new Uri($"https://{keyVaultName}.vault.azure.net/"),
                        new DefaultAzureCredential());
                    builder.ConfigurationBuilder.AddAzureKeyVault(secretClient, new KeyVaultSecretManager());
                }
            }

            _configuration = builder.ConfigurationBuilder.Build();
        }

**Comment by markgibbons25** [comment]
Thanks @UsmanMohammad , it's amazing that this is still an issue years later and not documented on the [DI page](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection).

**Comment by ghost** [comment]
This just totally ruined my morning. We had an odd issue on a production AZ Function, so i disabled a trigger so that I could attempt to process it locally and debug what was going on, yet still allowing the messages to queue up. Then all hell broke loose. All of a sudden, everything to do with those hosted APIs on the AZ function started redirecting to development assets. 

Not of fan of how this works.

**Comment by StefanosBekiaris** [comment]
Any updates on this? Currently setting the `AZURE_FUNCTIONS_ENVIRONMENT` always returns Development

**Comment by fabiocav** [comment]
Noticed this was still open without clear information from the PR linked, so I wanted to share the details:

- This issue has been addressed in 4.0, and the default behavior there matches the expectation. No action required
- If running in 3.0, the default behavior has not been changed,  you you can opt into the behavior that will address the issue by defining an App Setting named `AzureWebJobsFeatureFlags` with a value of `DisableDevModeInDebug`

Hope this information helps. Closing as resolved, but let us know if there are any additional questions. Thanks!
